### DISCUSSION TITLE
TCP keepalive on server side

### DISCUSSION THREAD

**Discussion by oldmanwombat** [FreeOpcUa/opcua-asyncio#396] [Q&A]
This is more of a question than an issue.

Is there currently a way to set TCP keepalive on the server socket?

I haven't tried it yet, but it [looks to be possible ](https://stackoverflow.com/questions/12248132/how-to-change-tcp-keepalive-timer-using-python-script) to do this after the socket is opened. But I think it would be nicer if it was integrated into the server config

Best regards,
Nick

edit 1:

So I've found the point where the server is being created.

[https://github.com/FreeOpcUa/opcua-asyncio/blob/cf278b594c99d9deb3729fb78ad4999cfe7031b5/asyncua/server/binary_server_asyncio.py#L129](https://github.com/FreeOpcUa/opcua-asyncio/blob/cf278b594c99d9deb3729fb78ad4999cfe7031b5/asyncua/server/binary_server_asyncio.py#L129)

I'm not sure how I can expose the socket (or if it's even possible). And if it is possible, is there a way to enable it like written [here](https://stackoverflow.com/questions/12248132/how-to-change-tcp-keepalive-timer-using-python-script). It's straight forward when using the built in socket library. 

edit 2:
Since the socket objects are available through the returned [Server object](https://docs.python.org/3/library/asyncio-eventloop.html?highlight=create_server#asyncio.Server.sockets), is it as simple as adding something like this just after the server is created?
```
self._server.sockets[0].setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
```

**Comment by oroulet** [comment]
how do you use that feature? What is your goal?

**Reply by oldmanwombat** [reply]
End goal is just to enable the tcp-keepalive on the server socket. I've been told this is a generally recommended thing for some servers, as it can help prevent denial of service events.

**Reply by AndreasHeine** [reply]
https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server
`coroutine loop.create_server(protocol_factory, host=None, port=None, *, family=socket.AF_UNSPEC, flags=socket.AI_PASSIVE, sock=None, backlog=100, ssl=None, reuse_address=None, reuse_port=None, ssl_handshake_timeout=None, start_serving=True)`

> backlog is the maximum number of queued connections passed to listen() (defaults to 100).

https://github.com/FreeOpcUa/opcua-asyncio/blob/cf278b594c99d9deb3729fb78ad4999cfe7031b5/asyncua/server/binary_server_asyncio.py#L129

so i would assume after 100 connections you have your "DoS"

would be neat feature do sort those inactive clients out aswell es a possible auto reconnect behaivior!?

**Reply by oldmanwombat** [reply]
Correct. This is what the tcp-keepalive will prevent, after a certain time it will check that the client is still connected and kick it off if there's no response.

**Reply by AndreasHeine** [reply]
unfortunatly i cant find that socket type `socket.SO_KEEPALIVE` looks like they renamed it `socket.SIO_KEEPALIVE_VALS`
https://docs.python.org/3/library/socket.html

i would assume that you can predifine the sock argument in create_server like:

define socket with self.hostname, self.port
`self._server = await self.iserver.loop.create_server(self._make_protocol, sock=_your predifined socket_)`


https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.create_server
`coroutine loop.create_server(protocol_factory, host=None, port=None, *, family=socket.AF_UNSPEC, flags=socket.AI_PASSIVE, sock=None, backlog=100, ssl=None, reuse_address=None, reuse_port=None, ssl_handshake_timeout=None, start_serving=True)`

> sock can optionally be specified in order to use a preexisting socket object. If specified, host and port must not be specified.

**Reply by oldmanwombat** [reply]
Sorry for the late reply, I had some other issues I had to solve first.

I think you're right, that's one way to do it. I done it slightly differently, the socket object is exposed through `self._server.sockets[0]` and you can enable the tcp-keepalive directly there.

I'll tidy it up a bit and make a PR, but just for reference here's the quick change I made and tested that it works.

https://github.com/nickjhall/opcua-asyncio/blob/caeb2579e22518e4ff4bd761e872401328e0b301/asyncua/server/binary_server_asyncio.py#L139
